File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let map = {}
    let left = 0

    return s.split('').reduce((max, val, idx) => {
        const lastSeenIdx = map[val]

        const alreadyIncluded = left <= lastSeenIdx

        left = alreadyIncluded ? lastSeenIdx + 1 : left

        map[val] = idx

        const currentWindowSize = idx - left + 1

        return Math.max(max, currentWindowSize)
    }, 0)
    
};

// sliding window
// map is {}
// left is 0
// s.split().reduce(max/acc, val/cur, idx)
// get last seen index from map
// if left is less than or equal to last seen index, the character is already included in the window
// if included in the window, left is last seen index + 1 : left
// update the value index in map
// current window size is idx - left + 1
// return max between current window size and max

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:      
        seen = []
        left = 0
        right = 0
        length = 0
        while right < len(s):
            if s[right] in seen:
                length = max(length, len(seen))
                left += 1
            else:
                seen.append(s[right])
            right += 1
        length = max(length, len(seen))
        return length